topn.c

```diff
@@ -473,6 +473,127 @@ topn_union_trans(PG_FUNCTION_ARGS)
 		/* No need to check the size again since it is already checked in MergeTopn */
 	}
 
+	PG_RETURN_POINTER(topnTrans);
+}
+
+
+/*
+ * topn_serialize function converts TopnAggState to bytea
+ */
+Datum
+topn_serialize(PG_FUNCTION_ARGS)
+{
+	Size topnArraySize;
+	TopnAggState *topnTrans = (TopnAggState *) PG_GETARG_POINTER(0);
+	HASH_SEQ_STATUS status;
+	FrequentTopnItem *currentTask = NULL;
+	bytea *ret;
+	char *bpPtr; /* Cursor for writing into ret */
+	long hashTableSize = hash_get_num_entries(topnHashtable(topnTrans));
+
+	/* it must be called as a transition routine or it fails */
+	if (!AggCheckCallContext(fcinfo, NULL))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_DATA_EXCEPTION),
+				 errmsg("topn_serialize outside transition context")));
+	}
+
+	topnArraySize = sizeof(FrequentTopnItem) * hashTableSize;
+	ret = palloc(VARHDRSZ + topnArraySize);
+	SET_VARSIZE(ret, VARHDRSZ + topnArraySize);
+	bpPtr = (void *) VARDATA(ret);
+
+	hash_seq_init(&status, topnHashtable(topnTrans));
+
+	while ((currentTask = (FrequentTopnItem *) hash_seq_search(&status)) != NULL)
+	{
+		memcpy(bpPtr, currentTask, sizeof(FrequentTopnItem));
```

Comment:
Minor note: It strikes me that a lot of these serialize functions might have more subtle over-the-network issues like the potential for different character encodings across servers. We can probably ignore that here.